libs/pubsub/pubsub.go

```diff
@@ -332,24 +336,34 @@ loop:
 }
 
 func (state *state) add(clientID string, q Query, ch chan<- interface{}) {
+	qStr := q.String()
 
 	// initialize clientToChannelMap per query if needed
-	if _, ok := state.queries[q]; !ok {
-		state.queries[q] = make(map[string]chan<- interface{})
+	if _, ok := state.queryToChanMap[qStr]; !ok {
+		state.queryToChanMap[qStr] = make(map[string]chan<- interface{})
 	}
 
 	// create subscription
-	state.queries[q][clientID] = ch
+	state.queryToChanMap[qStr][clientID] = ch
+
+	// initialize queries if needed
+	if _, ok := state.queries[qStr]; !ok {
+		state.queries[qStr] = &queryPlusRefCount{q: q, refCount: 0}
+	}
+	// increment reference counter
+	state.queries[qStr].refCount++
```

Comment:
I hope everyone touching this code will be as good of a coder as you and we will never forget to increment/decrement the `refCount` (as done in this pr) 🙈

Reply:
I know it looks a bit dangerous, but I believe the game is worth the candle. 

pros:
- we don't store a pointer to a query in pubsub#subscriptions ("preserve original query" hack)
- we don't store a pointer to a query if `state.queries[qStr]` already exists

cons:
- need to increment / decrement ref. counter

Note: we'll probably get rid of the pubsub in the future thanks to Zarko's suggestion. When/if this happens, I'd still like us to preserve the code, maybe even extract it to a separate repo cause I think the new design is especially good.